linkd

xla opened the pull request:
First iterative implementation of [RFC 0696][1] focusing on driving the core
peer protocol and establishing the harness for configuration of a running node.
Rite of passage for it was the replacement of the ephemeral peer in the e2e
networkss, as that required a well behaved peers with a set of knobs exposed.

Structurally the majority implementation lives in a library crate `linkd-lib` -
probably the most horrendous crate name in this repo to date - with a newly
created workspace under `bins` and a `linkd` crate in wrapping it lightly in a
`main`. That strategy allows for a checked-in `Cargo.lock` for binaries
associated with link project. Beyond that the code is rather Mario-esque as it
is primarily plumbing of code that either existed in other crates backing
binaries or newish code doing the same thing for core pieces.

There are still large pieces missing, which a tracked in [#722][2] and will be
piled on-top as patches to keep this already big delta focused. The declared
initial goal is to get `linkd` into a state where it can run as bootstrap/seed
node.

[1]: https://github.com/radicle-dev/radicle-link/blob/master/docs/rfc/0696-p2p-node.adoc
[2]: https://github.com/radicle-dev/radicle-link/issues/722

FintanH reviewed (commented):
Just some small things so far, but it's looking like it's coming together :)

xla commented on librad/src/profile.rs:
```diff
@@ -50,6 +52,15 @@ impl Default for RadHome {
     }
 }
 
+impl fmt::Display for RadHome {
```
Structopt requires a `Display` impl.

FintanH replied:
Sorry, I wasn't clear enough. I'm suggesting the following:
```rust
impl fmt::Display for RadHome {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self)
    }
}
```

So we're just setting the `Display` implementation to be the same as the `Debug` one, which is almost what you're doing here but with slight differences.

xla replied:
a03bd32

FintanH replied:
:heart:

FintanH commented on linkd/src/cfg/seed.rs:
```diff
@@ -0,0 +1,74 @@
+    /// # Errors
+    ///
+    /// * If the supplied address cannot be resolved.
+    async fn from_bootstrap(bootstrap: &Bootstrap) -> Result<Self, Error> {
```
```suggestion
    async fn try_from_bootstrap(bootstrap: &Bootstrap) -> Result<Self, Error> {
```

For idiotmaticness

xla replied:
It would be more idiomatic if the advent of async wouldn't been incompatible with a majority of language idioms...

485bfbd

FintanH replied:
That's why it's idiotmatic =]

FintanH commented on linkd/src/lib.rs:
```diff
@@ -0,0 +1,15 @@
+
+pub mod args;
+
+mod cfg;
```
bikeshed: we've used `config` or `Config` everywhere else, so I'd vote we stay consistent on the naming.

xla replied:
Can you point me to an approved RFC where we agreed on this? :bike: :derelict_house:

FintanH replied:
touché

FintanH commented on linkd/src/args.rs:
```diff
@@ -0,0 +1,216 @@
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Bootstrap {
+    pub addr: String,
```
Maybe not because these resolve to `SocketAddr`, right?

xla replied:
Don't think URL will serve us well, we don't use much of it otherwise.

FintanH replied:
Yup, fair :)

xla commented on bins/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
```
44bf404

xla commented on linkd/src/args.rs:
```diff
@@ -0,0 +1,216 @@
+    /// Home of the profile data, if not provided is read from the environment
+    /// and falls back to project dirs.
+    #[structopt(long, default_value, parse(from_str = parse_rad_home))]
+    pub rad_home: RadHome,
```
This takes precedence.

FintanH replied:
But if it's not specified then it defaults to `RadHome::ProjectDirs`, so I could do:
```
$ RAD_HOME=/tmp/rad-home linkd --ssh-agent
```
and then it would end up using the project dirs

xla replied:
We could employ structopts magic of reading env variables. I wonder tho, will this be hard to communicate what arg will have what value - the matrix of combinations is expanding. Furthermore the daemon is unlikely be run just from a users tty. Where and when is it usual to set the `RAD_HOME`?

FintanH replied:
> Where and when is it usual to set the RAD_HOME?

I would say that it should be used for testing or someone who _really_ wants their home to be somewhere else. So I think this parameter should just be removed. 

For its use here[[0]], I would recommend this usage[[1]].

[0]: https://github.com/radicle-dev/radicle-link/pull/779/files#diff-690b715b0168b0abda74382805e6b80dd3e8ffed9bd71b7b7739d281361ea959R125-R131
[1]: https://github.com/radicle-dev/radicle-link/blob/master/rad-profile/src/lib.rs#L72

xla replied:
I don't think I agree, if we remove this argument then in order to fully configure the daemon you are forced to employ a mi of env vars and cli options.

xla replied:
> For its use here[[0](https://github.com/radicle-dev/radicle-link/pull/779/files#diff-690b715b0168b0abda74382805e6b80dd3e8ffed9bd71b7b7739d281361ea959R125-R131)], I would recommend this usage[[1](https://github.com/radicle-dev/radicle-link/blob/master/rad-profile/src/lib.rs#L72)].

Sorry I was looking for places where the env var `RAD_HOME` is set.

FintanH replied:
> I don't think I agree, if we remove this argument then in order to fully configure the daemon you are forced to employ a mix of env vars and cli options.

I think in this particular case it's not going to be set that often. So you won't realistically be mixing env and cli options. 

> Sorry I was looking for places where the env var RAD_HOME is set.

Upstream's sandbox environment :shrug: It doesn't tend to be set for actual use.

xla replied:
> Upstream's sandbox environment shrug It doesn't tend to be set for actual use.

Reinforces my impression to try to weave in these env vars and always expect explicit configuration through the cli.

FintanH replied:
Then you need to find a way for this CLI parameter to play well with the env var

xla replied:
> Then you need to find a way for this CLI parameter to play well with the env var

If it's never set, why bother keeping it the env var around?

FintanH replied:
It doesn't mean it _can't_ be set. Like I said above, the following command would have undesirable effects:
```
$ RAD_HOME=/tmp/rad-home linkd --ssh-agent
```

xla replied:
1b60712

alexjg commented on linkd-lib/src/cfg.rs:
```diff
@@ -0,0 +1,183 @@
+    Base64(#[from] base64::DecodeError),
+
+    #[error(transparent)]
+    IO(#[from] io::Error),
```
nit: should be `Io`

alexjg commented on linkd-lib/src/socket_activation/unix.rs:
```diff
@@ -0,0 +1,67 @@
+const LISTEN_PID: &str = "LISTEN_PID";
+
+pub fn env() -> Result<Option<UnixListener>> {
+    // TODO(xla): Enable usage of more than the first fd. For now the assumption
```
What would be the purpose of using multiple file descriptors?

xla replied:
Systemd unit files support the declaration of N named sockets, these are then used by the binary at own discretion. Use-cases include but not limited to offering multiple APIs or offer telemetry on an isolated socket.

> sd_listen_fds_with_names() is like sd_listen_fds(), but optionally also returns an array of strings with identification names for the passed file descriptors

From https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html

alexjg replied:
Gotcha, thanks.

kim commented on linkd-lib/Cargo.toml:
```diff
@@ -0,0 +1,42 @@
+log                 = "0.4"
+nix                 = "0.22"
+signal-hook         = { version = "0.3", default-features = false }
+signal-hook-tokio   = { version = "0.3", features = [ "futures-v0_3" ] }
```
This does not support Windows, as it uses a "pipe" which actually is a tokio `UnixStream` for signal delivery. It saves a thread, yes, but I'm not sure this is overall cheaper.

xla replied:
Was mostly optimising for ergonomics as this fits better with the style of spawning a task per subroutine. While the failing windows tests haunted me in my dreams the last days I concluded it might be worthwhile porting the parked thread with atomics approach we have in other places over and give the impression that this project works on Windows after all. Maybe then @FintanH stops calling me out on it.

kim commented on linkd-lib/Cargo.toml:
```diff
@@ -0,0 +1,42 @@
+structopt           = { version = "0.3", default-features = false }
+thiserror           = "1.0"
+tempfile            = "3.2"
+tokio               = { version = "1.10", default-features = false, features = [ "fs", "io-std", "macros", "process", "rt-multi-thread", "signal" ] }
```
Do you need `signal`?

xla replied:
Maybe not, or maybe I drop the other signal related code in favour of this. What could go wrong, tokio is production ready software, right?

kim replied:
tokio only knows how to handle `CTRL-C` on windows (there's a dedicated API for that). Maybe that's enough, idk, how does one run services there anyway?

xla replied:
I concluded until we have more insights into how link is operate on gaming rigs break and ctrl-c is what can be expected.

c792721

kim commented on linkd-lib/src/args.rs:
```diff
@@ -0,0 +1,298 @@
+    /// Forces the creation of a temporary root for the local state, sould be
+    /// used for debug and testing only.
+    #[structopt(long)]
+    pub tmp_root: bool,
```
IIUC, setting this to `true` will render `rad_home` a no-op. Perhaps fold into one argument?

kim commented on linkd-lib/src/args.rs:
```diff
@@ -0,0 +1,298 @@
+#[derive(Debug, Eq, PartialEq, StructOpt)]
+pub enum KeySource {
+    File,
+    InMemory,
```
I was wondering: "whose memory?"

Maybe this should be called `Ephemeral`, `AdHoc`, `Generate`, or somesuch?

xla replied:
ac88165

kim commented on linkd-lib/src/socket_activation/macos.rs:
```diff
@@ -0,0 +1,12 @@
+use anyhow::Result;
+
+pub fn env() -> Result<Option<UnixListener>> {
+    todo!()
```
:shrimp:

xla replied:
:apple:

FintanH commented on linkd-lib/src/logging.rs:
```diff
@@ -0,0 +1,51 @@
+// Copyright © 2021 The Radicle Link Contributors
```
I was thinking, logging might be useful to the `rad` CLI as well, so wondering if we'd be interested in putting this in `rad-clib`?

xla replied:
Generally yes, but we should build a proper :diamond_shape_with_a_dot_inside: and move the shared code (signer things, logging, etc.) into its own crate, otherwise linkd would depend on rad, which strikes me as out.

cloudhead commented on Cargo.toml:
```diff
@@ -12,6 +12,7 @@ members = [
   "link-crypto",
   "link-git-protocol",
   "link-identities",
+  "linkd-lib",
```
`link-daemon`?

kim replied:
I liked the `blink`

xla replied:
In my understanding `linkd-lib` is adhering to our non-RFCed standard of crate naming, `linkd` being the prefix of the binary that consumes this crate (same as the `rad` crates for the cli).

For comedic purpose and artificial increase of barrier to entry for contributors I'm not opposed to naming it `dlink`.

kim replied:
> the prefix of the binary that consumes this crate

Interesting, this interpretation didn't occur to me. I'm thinking of `rad` as an anagram of `cli`, where `cli` is a bunch of commandline executables, not necessarily that they must be invoked by typing `rad` first (the fact that we said it will be one executable for starters is an implementation detail if you will). I'm also thinking of these prefixes as a navigational aid in a monorepo structure, while I discarded the idea to enforce complete separation -- I would expect that there will always be things which seem worthwhile to put in their own crate, so they can be depended on in multiple places.

So tl;dr I feel `linkd-` as a top-level name is too specific, because we may have other daemons. Maybe we can find a different prefix conveying daemon-ness? It could be `daemon-`, although speling...

FintanH replied:
> different prefix conveying daemon-ness?

How about `matt`? 

![image](https://user-images.githubusercontent.com/3503528/132823473-11d3d829-a1c4-4468-a60a-9230e4e97a0e.png)

xla replied:
3d4e38c

cloudhead commented on linkd-lib/src/args.rs:
```diff
@@ -0,0 +1,298 @@
+
+impl Default for KeySource {
+    fn default() -> Self {
+        Self::Stdin
```
I'd think `File` is a more sensible default, otherwise we block on stdin and it's confusing

cloudhead replied:
In fact not sure why you'd want stdin if you have file

xla replied:
Sometime in the past, someone in an outburst of productivity produced overnight a critical piece of infrastructure. To aid the network in bootstrapping and bring replication to the plebs, born was the SEED. As the creator believed in composability and drank from the firehose called the UNIX pipe, they made it so that you can pipe the most private of private things - the secret key -  into the SEEDs stdin. It was glorious, but many an operator lost their sleep over the breaking systemd services, a lot of bash was written in those days (dark times indeed). And because of that legacy we have support for STDIN.

In fact, I thought I never say that, it has proven handy for configuration of the whale containers. There you often share the same image for different peers, so you can't really store a single key on disk as they would share it. Changing the default to file is not an outrageous thought.

cloudhead replied:
Epic

cloudhead commented on linkd-lib/src/cfg.rs:
```diff
@@ -0,0 +1,183 @@
+lazy_static::lazy_static! {
+    /// Localhost binding to any available port, i.e. `127.0.0.1:0`.
+    pub static ref LOCALHOST_ANY: SocketAddr =
+        SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 0));
```
`([0,0,0,0], 0).into()`?

kim replied:
That’s not a good default, either. I’d rather mandate that the bind address must be specified explicitly, and provide shortcuts such as “any”, “localhost”

xla replied:
bd8ee1b

cloudhead commented on linkd-lib/src/cfg/seed.rs:
```diff
@@ -0,0 +1,74 @@
+    ///
+    /// * If the supplied address cannot be resolved.
+    async fn try_from_bootstrap(bootstrap: &Bootstrap) -> Result<Self, Error> {
+        if let Some(addr) = lookup_host(bootstrap.addr.clone()).await?.next() {
```
Why not `collect` so we can load multiple seeds here?

xla replied:
Because they would share the same `peer_id` and that is a violation of the protocol. But you are absolutely right that we should collect all addresses and associate them with the same peer, as the discovery expects a stream of `(PeerId, Vec<SocketAddr>)`.

xla replied:
9d993d9

cloudhead commented on linkd-lib/src/logging.rs:
```diff
@@ -0,0 +1,51 @@
+    if env_logger::builder().try_init().is_ok() {
+        if env::var("RUST_LOG").is_err() {
+            env::set_var("RUST_LOG", "debug");
+        }
```
Maybe:
```rust
                .with_env_filter(
                    EnvFilter::try_from_default_env()
                        .unwrap_or_else(|_| tracing_subscriber::EnvFilter::new("debug")),
                )
```

xla replied:
7cf3e34

cloudhead commented on linkd-lib/src/protocol.rs:
```diff
@@ -0,0 +1,72 @@
+            Ok(bound) => {
+                let (stop, run) = bound.accept(disco.clone().discover());
+                let run = run.fuse();
+                pin_mut!(run);
```
I hate that this is a thing

cloudhead replied:
:mage_man: :sparkles:

xla replied:
Unfortunately fixing async Rusts Pinning abstraction is out-of-scope for this PR. :<

xla commented on linkd-lib/Cargo.toml:
```diff
@@ -0,0 +1,42 @@
+futures             = "0.3"
+lazy_static         = "1.4"
+log                 = "0.4"
+nix                 = "0.22"
```
Not your kind of nix, boy!

xla commented on linkd-lib/src/node.rs:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    if let Some(_listener) = socket_activation::env()? {
+        // TODO(xla): Schedule listen loop.
```
:bacon: :fried_egg:

xla commented on linkd-lib/src/args.rs:
```diff
@@ -0,0 +1,298 @@
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+// TODO(xla): Expose discovery args.
```
With our rapid offboarding of this walled garden I'm uncertain where to track follow-up work. Will go over the TODOs and see if they are out-of-sync with the #722.

xla commented on linkd-lib/src/args.rs:
```diff
@@ -0,0 +1,298 @@
+impl Default for MetricsArgs {
+    fn default() -> Self {
+        Self {
+            provider: None,
```
You are right, I haven't found a way to enforce a heuristic a la: only expect X when Y is set.

FintanH replied:
Ya, I meant the way one would usually do this is by packing it as an enum variant, right?
```rust
enum Provider {
   Graphite { address: String }
}
```
but that doesn't fly if we want flattened args, so I guess we're between a rock and a hard place :) I wonder would it make sense to have the `graphite_addr` as `Option` as well?

xla replied:
Might be I'm holding the Structops wrong but that would conflict with the need to default to the local graphite instance if graphite is set as a metric provider.

FintanH replied:
My thinking is that if `metrics-provider` is set to `"graphite"` then you'll end up with `Some(MetricsProvider::Graphite)` and then the address will be `Some("localhost:8080")` (or whatever the port is). And on the flip-side, if `metrics-provider` is `None` then `graphite_addr` is also, rightfully, `None`. 

That being said this is not a big deal, and should not be a blocker :) Just musing about what's the best way to represent the data

xla replied:
I don't follow how you'd express that with the structopt primitives.

FintanH replied:
My suggestion is:
```rust
#[derive(Debug, Eq, PartialEq, StructOpt)]
pub struct MetricsArgs {
    /// Provider for metrics collection.
    #[structopt(long = "metrics-provider", name = "metrics-provider")]
    pub provider: Option<MetricsProvider>,

    /// Address of the graphite collector to send stats to.
    #[structopt(
        long,
        required_if("metrics-provider", "graphite")
    )]
    pub graphite_addr: Option<String>,
}

/* then the conversion here */
        let metrics = match args.metrics.provider {
            Some(args::MetricsProvider::Graphite) => Some(Metrics::Graphite(
                args.metrics
                    .graphite_addr
                    .unwrap_or("localhost:2003")
                    .to_socket_addrs()?
                    .next()
                    .unwrap(),
            )),
            None => None,
        };
```

Which is a minor improvement if any. It's just unfortunate that we can't express this better in `structopt`.

xla commented on linkd-lib/src/metrics.rs:
```diff
@@ -0,0 +1,6 @@
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+pub mod graphite;
```
Besides this code path it wouldn't guard any dependencies as this doesn't have any besides `std`. In my understanding that is usually why crate consumers would want feature flags to not be exposed to certain sub-trees in the cargo tree. Not sure where we generally stand on the matter as to what should and needs to be enabled with flags.

FintanH replied:
The `std` point is good :) Happy to not stick it behind a flag :ok_hand:

xla commented on linkd-lib/src/cfg/seed.rs:
```diff
@@ -0,0 +1,74 @@
+
+        for bootstrap in bootstraps.iter() {
+            resolved.push(Seed::try_from_bootstrap(bootstrap).await?);
+        }
```
To me that seems like a configuration error and should be surfaced asap. Main reason being is that if we just bubble it up as log line, it will become the typical noise in operational logs and operators will not be acting on it.

cloudhead replied:
well, DNS can fail temporarily though, but yeah up to you!

FintanH replied:
> well, DNS can fail temporarily though, but yeah up to you!

Perhaps a retry mechanism can be added in a future iteration?

xla commented on linkd-lib/src/args.rs:
```diff
@@ -0,0 +1,298 @@
+    #[structopt(
+        long = "key-file-path",
+        name = "key-file-path",
+        default_value = "",
```
e1f8531

FintanH reviewed (approved):
Nice work =] Looking forward to the new age of daemons

kim commented:
I would, however, appreciate a slightly more condensed commit history.

xla commented:
> I would, however, appreciate a slightly more condensed commit history.

I always squash before merge but leave that to the latest point to account for commits addressing feedback.
